### PR TITLE
Npm vs Yarn fixes

### PR DESCRIPTION
**Pull request by sschuberth**
Please have a look at the individual commit messages for the details.

### CODE REVIEW

File: analyzer/src/main/kotlin/managers/Npm.kt [hunk 1]
```diff
@@ -416,6 +309,115 @@ open class Npm(
         return modulesDir.takeIf { it.name == "node_modules" }
     }
 
+    /**
+     * Construct a [Package] by parsing its _package.json_ file and - if applicable - querying additional
+     * content via the `npm view` command. The result is a [Pair] with the raw identifier and the new package.
+     */
+    @Suppress("HttpUrlsUsage")
+    internal fun parsePackage(packageFile: File): Pair<String, Package> {
```
Comment (fviernau) [inline analyzer/src/main/kotlin/managers/Npm.kt:312]: Is this function just being moved or also modified?

Would it make sense to factor out moving of code to a separate commit for easier review?

**Review by fviernau — DISMISSED** [review]

File: analyzer/src/main/kotlin/managers/Npm.kt (hunk 1 above)
Comment (sschuberth) [reply analyzer/src/main/kotlin/managers/Npm.kt:312]: It was moved out of static context, but in the same run needed to be modified (code was factored out to ``) to do the right thing for Yarn. I considered splitting the move from the modifications, but could not figure out how in this case (without leaving things intermediately in a broken state).

Comment (sschuberth) [reply analyzer/src/main/kotlin/managers/Npm.kt:312]: Ok, I found a way to improve this.

File: analyzer/src/main/kotlin/managers/utils/NpmDependencyHandler.kt [hunk 2]
```diff
@@ -49,13 +49,13 @@ data class NpmModuleInfo(
 /**
  * A specialized [DependencyHandler] implementation for NPM.
  */
-class NpmDependencyHandler : DependencyHandler<NpmModuleInfo> {
+class NpmDependencyHandler(private val npm: Npm) : DependencyHandler<NpmModuleInfo> {
```
Comment (oheger-bosch) [inline analyzer/src/main/kotlin/managers/utils/NpmDependencyHandler.kt:52]: This bidirectional dependency between `Npm` and `NpmDependencyHandler` is a bit problematic IMHO. Is there a way to move functionality, so that the relation becomes unidirectional again? For instance, could the `parsePackage()` function (and maybe further helper functions) be moved to `NpmDependencyHandler`?

File: analyzer/src/main/kotlin/managers/Npm.kt [hunk 3]
```diff
@@ -71,12 +71,6 @@ import org.ossreviewtoolkit.utils.common.withoutPrefix
 import org.ossreviewtoolkit.utils.core.log
 import org.ossreviewtoolkit.utils.spdx.SpdxConstants
 
-object NpmCli : CommandLineTool {
```
Comment (oheger-bosch) [inline analyzer/src/main/kotlin/managers/Npm.kt:74]: Not sure whether it makes a big difference, but you could keep the objects implementing `CommandLineTool`, and instead of extending the `CommandLineTool` interface in `Npm`, introduce a function (or property) like `getCommandLineTool()` that returns the tool for this concrete implementation.

File: analyzer/src/funTest/kotlin/managers/NpmFunTest.kt [hunk 4]
```diff
@@ -44,7 +44,7 @@ class NpmFunTest : WordSpec() {
                 val workingDir = projectsDir.resolve("shrinkwrap")
```
Comment (oheger-bosch) [inline analyzer/src/funTest/kotlin/managers/NpmFunTest.kt:44]: Commit message nit: Should it be "adhere *to* Java naming conventions"?

File: analyzer/src/main/kotlin/managers/Npm.kt (hunk 3 above)
Comment (sschuberth) [reply analyzer/src/main/kotlin/managers/Npm.kt:74]: In the current architecture, it's actually more convenient if really the package manager classes themselves implement the `CommandLineTool` interface, as they then show up in the correct section for `ort requirements`. Also see my remark in [this commit message](https://github.com/oss-review-toolkit/ort/pull/5162/commits/0c8cd71e6ef741a3816fd17fb165c55ad27b5bec).

File: analyzer/src/main/kotlin/managers/utils/NpmDependencyHandler.kt (hunk 2 above)
Comment (sschuberth) [reply analyzer/src/main/kotlin/managers/utils/NpmDependencyHandler.kt:52]: Unfortunately, `parsePackage()` needs to call a package-manager-specific function. Currently, the only way I see to avoid this is to also introduce a `YarnDependencyHandler` which inherits from `NpmDependencyHandler`. But I'm not sure yet whether that's much nicer.

File: analyzer/src/main/kotlin/managers/Npm.kt (hunk 1 above)
Comment (fviernau) [reply analyzer/src/main/kotlin/managers/Npm.kt:312]: Thanks!

**Review by oheger-bosch — DISMISSED** [review]

**Review by oheger-bosch — APPROVED** [review]

**Comment by sschuberth** [comment]
Merging despite the unrelated `DotNetFunTest` socket timeout.
